PHP7: use uopz for mocking in unit tests

macbre opened the pull request:
[PLATFORM-2138](https://wikia-inc.atlassian.net/browse/PLATFORM-2138)

This PR introduces changes that are compatible with both PHP 5.6 and 7.
- `WikiaMockProxyUopz` class introduced that is used by PHP7 - it uses `uopz` for mocking and modifying the behavior of the `new` operator. This class will be merged with `WikiaMockProxy` once we fully migrate to PHP7.
- `WikiaDispatcher` now calls `uopz_get_mock` to allow mocking of Nirvana controllers - `uopz` does not support classes mocking when `new` operator is provided with a class name by a string (see 00702ad)

@wladekb / @artursitarski 

Pinging @nandy-andy  and @saipetch to review 10588ab (`MonetizationModuleHelper` changes to make mocking possible in PHP7 - dynamic and static methods were mixed there)

Commit:
Disable XDebug's code coverage when running unit tests

It was braking uopz_set_mock()

macbre committed:
WikiaMockProxyUopz: allow mocking in PHP5.6 and 7

macbre committed:
Mark unit tests for quick debugging

macbre committed:
WIP

Commit:
WikiaMockProxyUopz: make global functions mocking work

Commit:
Revert "WIP"

This reverts commit 16b5e8b30a70596422e5ac03e4ee14250396874a.

macbre committed:
VignetteRequestTest: fix DivisionByZeroError exception

macbre committed:
WikiaMockProxyUopz: be less noisy when mocking new operator

Commit:
Fix MemcachedStatsTest - use wfWikiId() directly

Commit:
Fix ForumTest and WikiaIFrameTagBuilderHelperTest

RuntimeException: failed to set return for WikiaIFrameTagBuilderHelper::isMobileSkin, the method is defined in WikiaTagBuilderHelper

Commit:
Fix WikiaQuizModuleTest

macbre committed:
Merge branch 'dev' into php7-uopz

macbre committed:
ExacctTarget: fix unit test for PHP7

Normalize resources to allow assertEquals to pass in PHP7

@mixth-sense

Commit:
DiffEngineTest: less strict assert on generated markup

macbre committed:
FounderProgressBarTest: always return the latest DB mock

Commit:
WikiaDispatcher: use uopz_get_mock to get mocked controllers when running unit tests

macbre commented on extensions/wikia/ExactTargetUpdates/tests/client/ExactTargetClientTest.php:
```diff
@@ -202,6 +202,10 @@ public function testDefaultClientConstruction() {
 
+		# normalize resources to allow assertEquals to pass in PHP7
+		$expected->sdl = is_resource( $expected->sdl );
+		$soapClient->sdl = is_resource( $soapClient->sdl );
```
Pinging @mixth-sense (touches code owned by @Wikia/sustaining-engineering-team)

mixth-sense replied:
👍 LG, tej - although this code is going to be retired completely in a matter of days.

macbre committed:
WikiaDispatcher: show the exception class when handling internal throw

macbre committed:
MonetizationModuleHelper: do not mix static and dynamic methods when mocking

PHPUnit complains with PHPUnit_Framework_MockObject_BadMethodCallException

Commit:
NewrelicDriverTest: improve mocking of not existing functions

macbre committed:
Fix MockNamespacedClassesTest

Commit:
WikiaMaps: fix "string to number casting" test case for PHP7

Commit:
CreateNewWikiControllerTest: mock GlobalTitle::newFromText static call

macbre commented on extensions/wikia/CreateNewWiki/tests/CreateNewWikiControllerTest.php:
```diff
@@ -74,7 +74,7 @@ public function testCreateWikiSuccess( $testCase, $testData ) {
 
 		$this->mockClass('CreateWiki', $createWiki);
-		$this->mockClass('GlobalTitle', $mainPageTitle);
+		$this->mockClass('GlobalTitle', $mainPageTitle, 'newFromText');
```
@Wikia/sustaining-engineering-team

mixth-sense replied:
👍 LG, tej

Commit:
CrossWikiCoreTest: fix mocking of wfMessage and Message

macbre committed:
Search/ConfigTest: do not mix static and dynamic methods mocks

Commit:
NewrelicDriverTest: create a dummy newrelic_custom_metric() before mocking it

Commit:
ImageServingIndexerTest: fix mocking of DatabaseMysqlBase::replace

macbre committed:
ImageServingIndexerTest: disable access to page props

Commit:
Merge branch 'dev' into php7-uopz

macbre committed:
ConfigureWikiFactory: mock wgCreateDatabaseActiveCluster

@mixth-sense

macbre commented on extensions/wikia/CreateNewWiki/tasks/ConfigureWikiFactory.php:
```diff
@@ -82,7 +81,7 @@ public function getStaticVariables( $siteName, $imagesURL, $imagesDir, $dbName,
 		}
 
-		wfGetLBFactory()->sectionsByDB[$dbName] = $wikiFactoryVariables['wgDBcluster'] = $wgCreateDatabaseActiveCluster;
+		wfGetLBFactory()->sectionsByDB[$dbName] = $wikiFactoryVariables['wgDBcluster'] = \F::app()->wg->CreateDatabaseActiveCluster;
```
Pinging @Wikia/sustaining-engineering-team

mixth-sense replied:
👍 LG, tej

macbre commented on extensions/wikia/CreateNewWiki/tests/tasks/ConfigureWikiFactoryTest.php:
```diff
@@ -17,6 +17,8 @@ public function setUp() {
 		);
 		$this->mockClass( 'TaskContext', $this->taskContextMock );
+
+		$this->mockGlobalVariable( 'wgCreateDatabaseActiveCluster', 'c7' );
```
Pinging @Wikia/sustaining-engineering-team

mixth-sense replied:
👍 LG, tej

nandy-andy commented:
@macbre from my perspective the changes in https://github.com/Wikia/app/commit/10588abfa14699de9794eb26622428d515dbb85f look good Thanks :+1:

Commit:
MockNamespacedFunctionsTest: fix a test for function mocks

WikiaMockProxy::parseGlobalFunctionName removes slashes for the begining of function names

mixth-sense commented on extensions/wikia/WikiaMaps/tests/WikiaMapsParserTagControllerTest.php:
```diff
@@ -64,7 +64,7 @@ public function validateParseTagParamsDataProvider() {
 				'Invalid latitude parameter',
 				[ 'map-id' => 1, 'lat' => '0x32' ],
-				true
+				version_compare(PHP_VERSION, '7..0.', '>=') ? false : true # PHP7 does not cast strings to numbers
```
`7..0.` - is this correct?

macbre replied:
Good catch, fixed in c4c4802

macbre committed:
Remove parseGlobalFunctionName from WikiaBaseTest

mixth-sense commented on includes/wikia/nirvana/WikiaDispatcher.class.php:
```diff
@@ -132,6 +132,15 @@ public function dispatch( WikiaApp $app, WikiaRequest $request ) {
 				$response->setTemplateEngine($controllerClassName::DEFAULT_TEMPLATE_ENGINE);
 
+				// uopz can not override classes returned by new operator when the class name is passed as a string
+				global $wgRunningUnitTests;
```
This is hacky as hell, tej :-)

wladekb replied:
https://github.com/krakjoe/uopz/issues/42 should help us avoid this hack

mixth-sense commented on includes/wikia/tests/core/WikiaMockProxy.class.php:
```diff
@@ -278,7 +283,12 @@ public function disable() {
 		$this->enabled = false;
-		unset_new_overload();
+
+		# TODO: remove when PHP7 migration is completed (PLATFORM-2138)
```
How will we find this when PHP7 migration is completed, tej?

macbre replied:
```
ack-grep "PHP7" --php | grep TODO
```

Seriously, there's a ticket for post-migration actions.

wladekb replied:
Sure thing, dude!

macbre committed:
WikiaMapsParserTagControllerTest: fix a typo

macbre committed:
AssetsManagerBaseBuilderTest: test JS minifiers (jsmin and YUI)

wladekb commented on extensions/wikia/WikiaMaps/tests/WikiaMapsParserTagControllerTest.php:
```diff
@@ -64,7 +64,7 @@ public function validateParseTagParamsDataProvider() {
 				'Invalid latitude parameter',
 				[ 'map-id' => 1, 'lat' => '0x32' ],
-				true
+				version_compare(PHP_VERSION, '7.0.0.', '>=') ? false : true # PHP7 does not cast strings to numbers
```
@macbre you meant only strings with hex numbers, didn't you?

macbre replied:
Yep - 437510edc93cda7b3

wladekb commented on includes/wikia/services/tests/DiffEngineTest.php:
```diff
@@ -13,6 +13,6 @@ public function testGetDiff() {
 
 		$diff = $engine->generateDiffBody( $textA, $textB );
-		$this->assertContains( '<span class="diffchange diffchange-inline">foo</span>', $diff );
+		$this->assertContains( 'class="diffchange diffchange-inline">foo</', $diff );
```
@macbre what's changed here?

macbre replied:
`<ins>` and `<del>` tags are used by `wikidiff2` extension version for PHP7

wladekb commented on includes/wikia/tests/core/WikiaBaseTest.class.php:
```diff
@@ -92,7 +92,7 @@ protected function setUp() {
 			throw new Exception("Previous test did not execute tearDown()");
 		}
-		$this->mockProxy = new WikiaMockProxy();
+		$this->mockProxy = function_exists('runkit_method_rename') ? new WikiaMockProxy() : new WikiaMockProxyUopz();
```
Imagine having php5.6 without runkit but with uopz enabled. I'd rather base this check on php version

macbre replied:
Used `version_compare` in 0e87368

wladekb commented:
overall looks promising, I left a few comments inlined

Commit:
WikiaBaseTest: use version_compare to choose between WikiaMockProxy and WikiaMockProxyUopz

macbre committed:
WikiaMapsParserTagControllerTest: extend a comment
